doc/cprover-manual/modeling-shadow-memory.md

```diff
@@ -0,0 +1,148 @@
+# Shadow Memory
+
+(The original artefact's implementation is outlined in the paper [CBMC-SSM:
+Bounded Model Checking of C Programs with Symbolic Shadow
+Memory](https://dl.acm.org/doi/abs/10.1145/3551349.3559523).)
+
+## Introduction
+
+CBMC contains a component that supports *Symbolic Shadow Memory*. Shadow memory
+allows one to create a parallel memory structure that is maintained by the
+analysis program and that shadows the standard memory of the program. In that
+memory structure, a user can create fields, for which he can later set and
+retrieve values.
+
+By doing so, one can organise his own tracking of metadata related to the code,
+in a way that is then considered by the backend of CBMC during analysis. This
+can be used to implement novel analyses that the CBMC framework itself doesn't
+support - for example, the paper above presents as an example a taint analysis
+implemented with the use of the SSM component described here.
+
+## Usage
+
+A user can interact with the Symbolic Shadow Memory component (from now on
+onwards, SSM) through four CBMC primitives, which allow one to declare shadow
+memory fields, and retrieve/set their corresponding values:
+
+* `__CPROVER_field_decl_local`
+* `__CPROVER_field_decl_global`
+* `__CPROVER_get_field`
+* `__CPROVER_set_field`
+
+More precisely, their signatures (in pseudo-C, because of some constraints that
+we can't express using the type system), along with some small examples of their
+usage, look like this:
+
+### `void __CPROVER_field_decl_local(type1 field_name, SSM_value_type init_value)`
+
+Type constraints:
+
+* `type1`: string literal, such as `"field"`,
+* `SSM_value_type`: value of maximum 8 bits, signed or unsigned or `_Bool`.
+
+Declares a local (working for variable declared in the current function scope)
```

Comment:
If a user wants to use shadow memory for a value which goes through a function return, should the shadow memory be declared as local or global? It could be local based on it being on the stack or global based on it exceeding the lifetime of the function, so as a user I would be unsure which to declare it as.

Reply:
Interesting point.

I will explain better on a specific Note.

Reply:
So does that mean the shadow memory would have to be declared as global in that case?

Reply:
Nope, that is unfortunately a shortcoming of the actual implementation.

Global is for global variables and heap-allocated pointers.
Local is for the local function scope.

There may be in the future a way to propagate local shadow memory information to called functions and through return, but it is not implemented yet.

Reply:
Can we add the limitation to the documentation then? Do similar limitations apply to function arguments? Do they only apply to stack memory or can the limitations be dodged by heap allocating and passing pointers?

Reply:
Yes, it is possible to heap-allocate arguments and return types so they will be tracked on the global SM.

About the documentation I'm not too keen at adding a "limitation" section as it draws attention to the limitation itself.

As the documentation says already local is for local only and is not passed to sub-call, I added that the same applies to returns, so the only way around is having the passed-around values on the global SM.

Reply:
I disagree on the subject of documenting limitations. If we document them, then it allows users to work around them and allows us to point to the documentation if anyone runs across them. It also adds strength to the argument that removing such limitations is a feature request, rather than a bug fix. However I am happy to leave this to your discretion.

If using global shadow memory enables support for function arguments and returns then doesn't that answer my original question of whether local or global declarations are needed for that use case?